### PR TITLE
MedCat [#101]: extract biomedical concepts/entities from (free) text

### PR DESCRIPTION
**Pull request by Imipenem**
**PR Checklist**

-   [x] This comment contains a description of changes (with reason)
-   [x] Referenced issue is linked
-   [ ] If you've fixed a bug or added code that should be tested, add tests!
-   [x] Documentation in `docs` is updated

**Description of changes**

- MedCat object now only serves for cdb and vocab (not for actual processing functions)

- processing functions are now part of the public API rather than static object methods

- text annotation returns a DataFrame as a base for further analysis

- basic overview table exposed via API (with optional csv save)

- prep for adding to anndata object

- updated tsne, umap,pca and scatter plot to color by extracted entities

- paga, draw_graph, ... WIP

### CODE REVIEW

**Comment by Imipenem** [comment]
@Zethson : Already few things to discuss (mostly minor design things though).

For now, its implemented up to the annotation results.

File: ehrapy/plot/plot_medcat.py [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+    if status != "Both":
+        value_counts = medcat_results[
+            (medcat_results["meta_anns"] == "Affirmed") & (medcat_results["type_ids"] == tuis)
+        ]["pretty_name"].value_counts()
+    else:
+        value_counts = medcat_results[medcat_results["type_ids"] == tuis]["pretty_name"].value_counts()
+    # plot
+    sns.set_theme(style="whitegrid")
+    sns.barplot(x=value_counts.head(n), y=value_counts.head(n).index, orient="h", color="b")
+    plt.show()
```
Comment (Zethson) [inline ehrapy/plot/plot_medcat.py:28]: 1. Plots need to make use of the ehrapy settings object as much as possible. It defines matplotlib defaults etc. Pass these here, please.
2. It must be possible to pass or return plot axes.

-> scanpy plotting:
```
      show = settings.autoshow if show is None else show
        if not show:
            return dp.get_axes()
```

File: ehrapy/tools/nlp/_medcat.py [hunk 2]
```diff
@@ -156,213 +46,313 @@ def update_cat_config(self, concept_db_config: Config) -> None:
...
-        Returns:
-            A spacy Doc instance. Extract the entities using. doc.ents
-        """
-        return self.cat(text)
-
-    def print_cui(self, doc: Doc, table: bool = False) -> None:
-        """Prints the concept unique identifier for all entities.
+    def set_filter_by_tui(self, tuis: list[str] | None = None):
+        """Restrict results of annotation step to certain tui's (type unique identifiers).
+        Note that this will change the MedCat object by updating the concept database config. In every annotation
```
Comment (Zethson) [inline ehrapy/tools/nlp/_medcat.py:51]: 1 empty line after the summary. Always.
```suggestion

        Note that this will change the MedCat object by updating the concept database config. In every annotation
```

File: ehrapy/tools/nlp/_medcat.py [hunk 3]
```diff
@@ -156,213 +46,313 @@ def update_cat_config(self, concept_db_config: Config) -> None:
...
-            pass
-        else:
-            for entity in doc.ents:
-                # TODO make me pretty by ensuring that the lengths before the - token are aligned
-                print(f"[bold blue]{entity} - {entity._.cui}")
-
-    def print_semantic_type(self, doc: Doc, table: bool = False) -> None:
-        """Prints the semantic types for all entities.
+        As an exmaple:
+        Setting tuis=["T047", "T048"] will only annotate concepts (identified by a CUI (concept unique identifier)) in UMLS that are either diseases or
```
Comment (Zethson) [inline ehrapy/tools/nlp/_medcat.py:56]: Can we link to some documentation here to see which TUIs are available to select from?

File: ehrapy/tools/nlp/_medcat.py [hunk 4]
```diff
@@ -156,213 +46,313 @@ def update_cat_config(self, concept_db_config: Config) -> None:
...
+            tuis: list of TUI's
+        Returns:
 
-        Args:
-            text: Pandas Series of text to annotate.
-            print_statistics: Whether to print training statistics after training.
         """
-        print(f"[bold blue]Training using {len(text)} documents")
-        self.cat.train(text.values, progress_print=100)
+        if tuis is None:
```
Comment (Zethson) [inline ehrapy/tools/nlp/_medcat.py:64]: Do we really want this default filtering? Why is the default not exposed as a default parameter but rather hidden in the code?

I know that it takes ages if we don't do this filtering, but IMO the code must be more transparent then.

File: ehrapy/tools/nlp/_medcat.py [hunk 5]
```diff
@@ -156,213 +46,313 @@ def update_cat_config(self, concept_db_config: Config) -> None:
...
+    cdb.save(output_path)
+
+
+def load_concept_db(concept_db_path) -> CDB:
+    """Loads the concept database.
+
+    Args:
+        concept_db_path: Path to load the concept database from.
+    """
+
```
Comment (Zethson) [inline ehrapy/tools/nlp/_medcat.py:151]:
```suggestion
```

File: ehrapy/tools/nlp/_medcat.py [hunk 6]
```diff
@@ -156,213 +46,313 @@ def update_cat_config(self, concept_db_config: Config) -> None:
...
+def save_model_pack(ep_cat: MedCAT, model_pack_dir: str = ".", name: str = "ehrapy_medcat_model_pack") -> None:
+    """Saves a MedCAT model pack.
+
+    Args:
+        ep_cat: ehrapy's custom MedCAT object whose model should be saved
+        model_pack_dir: Path to save the model to (defaults to current working directory).
+        name: Name of the new model pack
+    """
+    # TODO Pathing is weird here (home/myname/...) will fo example create dir myname inside home inside the cwd instead of using the path
```
Comment (Zethson) [inline ehrapy/tools/nlp/_medcat.py:163]: Issue opened @Medcat?

File: ehrapy/tools/nlp/_medcat.py [hunk 7]
```diff
@@ -156,213 +46,313 @@ def update_cat_config(self, concept_db_config: Config) -> None:
...
+    _ = ep_cat.cat.create_model_pack(name)
+
+
+def run_unsupervised_training(
+    ep_cat: MedCAT, text: pd.Series, progress_print: int = 100, print_statistics: bool = False
```
Comment (Zethson) [inline ehrapy/tools/nlp/_medcat.py:168]: Naming of `ep_cat`

File: ehrapy/tools/nlp/_medcat.py [hunk 8]
```diff
@@ -156,213 +46,313 @@ def update_cat_config(self, concept_db_config: Config) -> None:
...
+    """
+    print(f"[bold blue]Running unsupervised training using {len(text)} documents.")
+    ep_cat.cat.train(text.values, progress_print=progress_print)
+
+    if print_statistics:
+        ep_cat.cat.cdb.print_stats()
+
+
+def annotate_text(
+    ep_cat: MedCAT, obs: pd.DataFrame, text_column: str, n_proc: int = 2, batch_size_chars: int = 500000
```
Comment (Zethson) [inline ehrapy/tools/nlp/_medcat.py:186]: Naming.

File: ehrapy/tools/nlp/_medcat.py [hunk 9]
```diff
@@ -156,213 +46,313 @@ def update_cat_config(self, concept_db_config: Config) -> None:
...
+    # note for overview, only one TUI and type is shown (there shouldn't be much situations were multiple are even possible or useful)
+    res = grouped.agg(
+        {
+            "pretty_name": (lambda x: next(iter(set(x)))),
+            "type_ids": (lambda x: next(iter(x))[0]),
+            "types": (lambda x: next(iter(x))[0]),
+            "row_nr": "nunique",
+        }
+    )
+    res = res.rename(columns={"row_nr": "n_patients"})
```
Comment (Zethson) [inline ehrapy/tools/nlp/_medcat.py:248]: Think that by default we take a patient_visit view. Might be a better naming here.

File: ehrapy/tools/nlp/_medcat.py [hunk 10]
```diff
@@ -156,213 +46,313 @@ def update_cat_config(self, concept_db_config: Config) -> None:
...
+    if save_to_csv:
+        res.to_csv(".")
+
+    overview_table = _df_to_rich_table(res.nlargest(n, "n_patients"))
+    console = Console()
+    console.print(overview_table)
+
+
+def _add_binary_column_to_obs(ep_cat: MedCAT, adata: AnnData, name: str) -> None:
+    """Adds a binary column to obs (temporarily) for plotting infos extracted from freetext.
```
Comment (Zethson) [inline ehrapy/tools/nlp/_medcat.py:263]:
```suggestion
    """Adds a binary column to obs (temporarily) for plotting infos extracted from freetext.
   
```

File: ehrapy/tools/nlp/_medcat.py [hunk 11]
```diff
@@ -156,213 +46,313 @@ def update_cat_config(self, concept_db_config: Config) -> None:
...
+    """Turn the flattened annotated results into a pandas DataFrame and remove duplicates."""
+    df = pd.DataFrame.from_dict(flattened_results, orient="index")
+    # remove duplicate entries; for example when a single entity like a disease is mentioned multiple times without any meaningful context changes
+    # Example: The patient suffers from Diabetes. Cause of the Diabetes, he receives drug X.
+    df.drop_duplicates(subset=["cui", "row_nr", "meta_anns"])
+    return df
+
+
+def _flatten_annotated_results(annotation_results: dict) -> dict:
+    """Flattens the nested set (usually 5 level nested) of annotation results.
```
Comment (Zethson) [inline ehrapy/tools/nlp/_medcat.py:283]:
```suggestion
    """Flattens the nested set (usually 5 level nested) of annotation results.
    
```

File: ehrapy/tools/nlp/_medcat.py [hunk 12]
```diff
@@ -156,213 +46,313 @@ def update_cat_config(self, concept_db_config: Config) -> None:
...
+    annotation_results is just a simple flattened dict with infos on all entities found
+    """
+    flattened_annotated_dict = {}
+    entry_nr = 0
+
+    # row numbers where the text column is located in the original data
+    for row_id in annotation_results.keys():
+        # all entities extracted from a given row
+        entities = annotation_results[row_id]["entities"]
+        # iterate over all entities extracted from the specific row
```
Comment (Zethson) [inline ehrapy/tools/nlp/_medcat.py:293]:
```suggestion
```
Useless comment.

File: ehrapy/tools/nlp/_medcat.py [hunk 13]
```diff
@@ -156,213 +46,313 @@ def update_cat_config(self, concept_db_config: Config) -> None:
...
+                        single_entity[entity_key] = entities[entity_id][entity_key]
+                    elif entity_key == "meta_anns":
+                        single_entity[entity_key] = entities[entity_id][entity_key]["Status"]["value"]
+                flattened_annotated_dict[entry_nr] = single_entity
+                entry_nr += 1
+    return flattened_annotated_dict
+
+
+def _format_df_column(df: pd.DataFrame, column_name: str) -> list[tuple[int, str]]:
+    """Format the df to match: formatted_data = [(row_id, row_text), (row_id, row_text), ...]
```
Comment (Zethson) [inline ehrapy/tools/nlp/_medcat.py:311]: comment style...

Maybe I have to turn on the docstring linting to annoy you haha :)

**Review by Zethson — CHANGES_REQUESTED** [review]
I'll have to see it in action to make a better assessment.

Maybe we need a quick call. As mentioned, I would like to see all of these function scoped into a class. IMO it makes sense to call these functions as ep.tl.mc.MEDCAT FUNCTION

The plots as well. ep.pl.mc.SOMETHING

File: ehrapy/tools/nlp/_medcat.py (hunk 3 above)
Comment (Imipenem) [reply ehrapy/tools/nlp/_medcat.py:56]: Line above has a link to the available TUIs

Comment (Zethson) [reply ehrapy/tools/nlp/_medcat.py:56]: Still a typo here though :)

"exmaple"

File: ehrapy/tools/nlp/_medcat.py (hunk 4 above)
Comment (Imipenem) [reply ehrapy/tools/nlp/_medcat.py:64]: Mutable types are not allowed (or at least not recommended) in python (as list would be). But we can discuss whether to remove this line completly of course. Might be more intuitive

File: ehrapy/tools/nlp/_medcat.py (hunk 13 above)
Comment (Imipenem) [reply ehrapy/tools/nlp/_medcat.py]: we should haha

File: docs/usage/usage.md [hunk 14]
```diff
@@ -229,7 +229,7 @@ In contrast to a preprocessing function, a tool usually adds an easily interpret
 
 ## Plotting
 
-The plotting module ehrapy.p**l largely parallels the tl.\* and a few of the pp.\* functions.
+The plotting module ehrapy.p\*\*l largely parallels the tl.\* and a few of the pp.\* functions.
```
Comment (Zethson) [inline docs/usage/usage.md:232]: Huh, where is this change coming from?

Comment (Imipenem) [reply docs/usage/usage.md:232]: Tbh, idk. Just ran nox and somehow this appeared.

Comment (Zethson) [reply docs/usage/usage.md:232]: Yeah, I figured this. We should examine whether this still renders properly

Comment (Imipenem) [reply docs/usage/usage.md:232]: renders incorrect, will update this

**Comment by Imipenem** [comment]
> 

@Zethson I had this in mind as well, but I'm not sure whether its working because rich import will be sorted by our pre-commit CI AFTER medcat import, so I'm not sure whether rich is always available or not. Might just have to try.

File: ehrapy/tools/nlp/_medcat.py [hunk 15]
```diff
@@ -93,276 +114,257 @@ def create_concept_db(csv_path: list[str], config: Config = None) -> CDB:
...
+        """
+        flattened_annotated_dict = {}
+        entry_nr = 0
+
+        # row numbers where the text column is located in the original data
+        for row_id in annotation_results.keys():
+            # all entities extracted from a given row
+            entities = annotation_results[row_id]["entities"]
+            for entity_id in entities.keys():
+                # ignore tokens for now
```
Comment (Zethson) [inline ehrapy/tools/nlp/_medcat.py:307]: Why? The good comments explain not what is happening, but why it is happening.

File: ehrapy/tools/nlp/_medcat.py [hunk 16]
```diff
@@ -93,276 +114,257 @@ def create_concept_db(csv_path: list[str], config: Config = None) -> CDB:
...
 
-        return AnnotationResult(results, patient_to_entities, cui_location, type_ids_location)
+    @staticmethod
+    def _filter_df_by_status(df: pd.DataFrame, status: str) -> pd.DataFrame:
+        """Util function to filter passed dataframe by status."""
+        df_res = df
+        if status != "Both":
+            if status not in {"Affirmed", "Other"}:
+                raise StatusNotSupportedError(f"{status} is not available. Please use either Affirmed, Other or Both!")
+            mask = df["meta_anns"].values == status
```
Comment (Zethson) [inline ehrapy/tools/nlp/_medcat.py:345]: Fine with me not to explicitly define the mask here and to simply just use it

**Review by Zethson — CHANGES_REQUESTED** [review]
Getting closer. I'll really need to see the adapted tutorial.

:+1:

File: ehrapy/tools/nlp/_medcat.py [hunk 17]
```diff
@@ -1,53 +1,74 @@
...
-    entities_pretty: dict[str, list[str]]
-    cui_locations: dict | None
-    type_ids_location: dict | None
+try:
+    from medcat.cat import CAT
+    from medcat.cdb import CDB
+    from medcat.cdb_maker import CDBMaker
+    from medcat.config import Config
+    from medcat.vocab import Vocab
+except ModuleNotFoundError:
```
[the remark this answers is not in the document]
Comment (Imipenem) [reply ehrapy/tools/nlp/_medcat.py:12]: The issue why I introduced this in the first place was that some tests were failing for unknown reasons because medcat has not been found (I double checked in the env and verified import was working in the REPL).
Maybe this is due to medcat being an extra dependency (which we should maybe reconsider).

File: ehrapy/tools/nlp/_medcat.py [hunk 18]
```diff
@@ -93,276 +114,257 @@ def create_concept_db(csv_path: list[str], config: Config = None) -> CDB:
...
         Args:
-            doc: A spacy tokens Doc.
-            table: Whether to print a Rich table.
+            model_pack_dir: Path to save the model to (defaults to current working directory).
+            name: Name of the new model pack
         """
-        # TODO implement Rich table
-        for ent in doc.ents:
-            print(ent, " - ", self.concept_db.cui2type_ids.get(ent._.cui))
+        # TODO Pathing is weird here (home/myname/...) will fo example create dir myname inside home inside the cwd instead of using the path
```
[the remark this answers is not in the document]
Comment (Imipenem) [reply ehrapy/tools/nlp/_medcat.py:163]: Will do ASAP.

File: ehrapy/tools/nlp/_medcat.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (Zethson) [reply ehrapy/tools/nlp/_medcat.py:12]: We should be able to make this work. medcat should stay an extra dependency imo because it has `transformers` as a dependency which then pulls tensorflow and jax. This can get messy very quickly.

File: ehrapy/tools/nlp/_medcat.py (hunk 18 above)
[the remark this answers is not in the document]
Comment (Imipenem) [reply ehrapy/tools/nlp/_medcat.py:169]: https://github.com/CogStack/MedCAT/issues/237
